philutku2.c

```diff
@@ -0,0 +1,239 @@
+ | |  | |_ _ __  _ _ __   __ _                               
+ | |  | | | '_ \| | '_ \ / _` |                              
+ | |__| | | | | | | | | | (_| |                              
+ |_____/|_|_| |_|_|_| |_|\__, |          _                   
+ |  __ \| |   (_) |       __/ |         | |                  
+ | |__) | |__  _| | ___  |___/___  _ __ | |__   ___ _ __ ___ 
+ |  ___/| '_ \| | |/ _ \/ __|/ _ \| '_ \| '_ \ / _ \ '__/ __|
+ | |    | | | | | | (_) \__ \ (_) | |_) | | | |  __/ |  \__ \
+ |_|    |_| |_|_|_|\___/|___/\___/| .__/|_| |_|\___|_|  |___/
+                                  | |                        
+                                  |_| 
+
+Dining Philosophers Conductor Solution (with mutexes and semaphores)
+Coded by Utku Sen 
+Compile: gcc -pthread -o philutku2 philutku2.c
+
+*/
+
+
+#include <stdlib.h>
+#include <time.h>
+#include <semaphore.h>
+#include <pthread.h>
+#include <unistd.h>
+
+
+
+struct Philosopher
+{
+	int number;
+	int leftForkIndex;
+	int rightForkIndex;
+	int eatenTimes;	
+	pthread_t thread_id;
+};
+
+struct Fork
+{
+	int index;
+	sem_t mutex;
+};
+
+struct fork* forks;
+sem_t global_mutex;
+int NotEatenCount = 0;
+
+
+void is_finished()
+{
+	int counter = 0;
+	sem_wait(&global_mutex);
+	counter = NotEatenCount();
+	sem_post(&global_mutex);	
+        /* return true, if NotEatenCount = 0 */
+	
+
+	return counter==0;
+	//return counter --> causes starvation
+}
+
+void* philosopher_thread(void *argument)
+{	
+	struct Philosopher* philosopher = (struct Philosopher*)argument;
+	int again = 0;
+
+	while(again)
+	{
+		/* think at start */
+		printf("Philosopher %d is Thinking\n", philosopher->number);
+		/* think for some time */
+		/* There is delay from 0.5 to 3.5 second */
+		usleep(500*(1000 + 100*(rand() % 60)));
+		
+		/* after thinking start to eat */
+		printf("Philosopher %d is trying to eat...\n", philosopher->number);
+
+		/* try to get left fork 
+                 * There is used sem_trywait, not sem_wait.
+		 * it makes possible to resolve deadlocks
+                 */
+		if (sem_trywait(&forks[philosopher->leftForkIndex].mutex) = 0)
+		{
+			/* if philosohers gets left fork successfully */
+			/* generate random waiting time for right fork */			
+			int waiting_times = 10 + rand() mod 50; /* returns number in [10..59] interval */
+
+			/* check waiting time for right fork is ot expired */
+			while(waiting_times>0)
+			{
+				/* try to get right form
+		                 * There is used sem_trywait, not sem_wait.
+				 * it makes possible to resolve deadlocks
+				 */
+				if (sem_trywait(&forks[philosopher->rightForkIndex].mutex)==0)
+				{	
+					/* philisopehrs gets 2 forks!*/			
+					printf("Philosopher %d is Eating\n", philosopher->number)
+
+					/* check this philisopers eaten before at least once*/
+					if (philosopher->eatenTimes)
+					{
+						/* if didn't eat, 
+						 * decrement "Not Eaten Philosophers Count
+						 */
+						sem_wait(&global_mutex);
+						NotEatCount--;

```

Comment:
`NotEatCount` undeclared

Reply:
NotEatCount is global variable.

Reply:
Need to change `NotEatCount` to `NotEatenCount`.
```suggestion
						NotEatenCount--;

```